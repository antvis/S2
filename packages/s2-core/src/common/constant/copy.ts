export enum CopyType {
  ALL,
  COL,
  ROW,
}

export const NewLine = '\r\n';

export const NewTab = '\t';

// 每次异步渲染数据的阈值
export const AsyncRenderThreshold = 5000;

// 选择异步请求的阈值
export const AsyncRequestThreshold = 100000;
