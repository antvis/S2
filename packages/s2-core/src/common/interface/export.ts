import type { SpreadSheet } from '../../sheet-type';
import type { DataItem, CellMeta } from '..';
import { EXTRA_FIELD } from '..';

export type MatrixPlainTransformer = (
  data: DataItem[][],
  separator?: string,
) => CopyablePlain;

export type MatrixHTMLTransformer = (data: DataItem[][]) => CopyableHTML;

export enum CopyMIMEType {
  PLAIN = 'text/plain',
  HTML = 'text/html',
}

export type CopyableItem = {
  type: CopyMIMEType;
  content: string;
};

export type CopyablePlain = {
  type: CopyMIMEType.PLAIN;
  content: string;
};

export type CopyableHTML = {
  type: CopyMIMEType.HTML;
  content: string;
};

export type CopyableList = [CopyablePlain, CopyableHTML];

export type Copyable = CopyableItem | CopyableItem[];

export type FormatOptions =
  | boolean
  | {
      formatHeader?: boolean;
      formatData?: boolean;
    };

export interface Transformer {
  [CopyMIMEType.PLAIN]: MatrixPlainTransformer;
  [CopyMIMEType.HTML]: MatrixHTMLTransformer;
}

export interface CopyOrExportConfig {
  selectedCells?: CellMeta[];
  formatOptions?: FormatOptions;
  separator?: string;
  customTransformer?: (transformer: Transformer) => Partial<Transformer>;
  isAsyncExport?: boolean;
}

export interface CopyAndExportUnifyConfig {
  separator: string;
  formatHeader: boolean;
  formatData: boolean;
  selectedCells: CellMeta[];
  transformers: Transformer;
  isAsyncExport: boolean;
}

export interface CopyAllDataParams {
  sheetInstance: SpreadSheet;
  split?: string;
  formatOptions?: FormatOptions;
  customTransformer?: (transformer: Transformer) => Partial<Transformer>;
  /** 是否开启异步导出 */
  isAsyncExport?: boolean;
}

export interface SheetCopyConstructorParams {
  spreadsheet: SpreadSheet;
  config: CopyOrExportConfig;
  isExport?: boolean;
}

export type MeasureQuery =
  | { [EXTRA_FIELD]: string | undefined }
  // eslint-disable-next-line @typescript-eslint/ban-types
  | {};
