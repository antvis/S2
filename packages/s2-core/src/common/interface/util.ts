export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
