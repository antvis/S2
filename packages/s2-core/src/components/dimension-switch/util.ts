import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
