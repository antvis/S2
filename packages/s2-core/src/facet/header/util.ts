import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
