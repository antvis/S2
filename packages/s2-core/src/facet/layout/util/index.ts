export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
