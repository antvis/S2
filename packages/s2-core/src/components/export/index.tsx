import React from 'react';
import { Menu, Dropdown, message } from 'antd';
import cx from 'classnames';
import { DotIcon } from '../icons/index';
import { copyData, copyToClipboard, download } from '../../utils/export';
import { SpreadSheet } from '@/sheet-type';
import { S2_PREFIX_CLS } from '@/common/constant/classnames';

export interface DataSet {
  icon?: React.ReactNode;
  name: string;
  value: string;
  type?: 'text' | 'location' | 'date';
  disabled?: boolean;
}

export interface ExportCfgProps {
  open: boolean;
  style?: React.CSSProperties;
  className?: string;
  icon?: React.ReactNode;
  copyOriginalText?: string;
  copyFormatText?: string;
  downloadOriginalText?: string;
  downloadFormatText?: string;
  successText?: string;
  errorText?: string;
  fileName?: string;
}
export interface ExportProps extends ExportCfgProps {
  sheet: SpreadSheet;
}

export const Export: React.FC<ExportProps> = ({
  className,
  icon,
  copyOriginalText = '复制原始数据',
  copyFormatText = '复制格式化数据',
  downloadOriginalText = '下载原始数据',
  downloadFormatText = '下载格式化数据',
  successText = '操作成功',
  errorText = '操作失败',
  sheet,
  fileName = 'sheet',
  ...restProps
}) => {
  const PRE_CLASS = `${S2_PREFIX_CLS}-export`;

  const exportData = (isFormat: boolean) => {
    const data = copyData(sheet, '\t', isFormat);
    if (copyToClipboard(data)) {
      message.success(successText);
    } else {
      message.error(errorText);
    }
  };

  const downLoadData = (isFormat: boolean) => {
    const data = copyData(sheet, ',', isFormat);
    try {
      download(data, fileName);
      message.success(successText);
    } catch (err) {
      message.error(errorText);
    }
  };

  const menu = (
    <Menu>
      <Menu.Item key="copyOriginal" onClick={() => exportData(false)}>
        <a href="`javascript:void(0)`">{copyOriginalText}</a>
      </Menu.Item>
      <Menu.Item key="copyFormat" onClick={() => exportData(true)}>
        <a href="`javascript:void(0)`">{copyFormatText}</a>
      </Menu.Item>
      <Menu.Item key="downloadOriginal" onClick={() => downLoadData(false)}>
        <a href="`javascript:void(0)`">{downloadOriginalText}</a>
      </Menu.Item>
      <Menu.Item key="downloadFormat" onClick={() => downLoadData(true)}>
        <a href="`javascript:void(0)`">{downloadFormatText}</a>
      </Menu.Item>
    </Menu>
  );

  return (
    <Dropdown
      overlay={menu}
      trigger={['click']}
      className={cx(PRE_CLASS, className)}
      {...restProps}
    >
      <a
        className="ant-dropdown-link"
        key="export"
        onClick={(e) => e.preventDefault()}
      >
        {icon || <DotIcon />}
      </a>
    </Dropdown>
  );
};
