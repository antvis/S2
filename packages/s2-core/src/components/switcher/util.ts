import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
