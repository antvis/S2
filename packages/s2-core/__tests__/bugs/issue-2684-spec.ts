/**
 * 透视表维值含有 "-", 复制数据时表头缺失
 * @description spec for issue #2684
 * https://github.com/antvis/S2/issues/2684
 */

import { isEmpty } from 'lodash';
import { InteractionStateName } from '../../src';
import { getSelectedData } from '../../src/utils/export/copy';
import * as mockDataConfig from '../data/data-issue-2684.json';
import { createPivotSheet } from '../util/helpers';
import type { S2Options, SpreadSheet } from '@/index';

const s2Options: S2Options = {
  width: 600,
  height: 480,
  interaction: {
    copy: {
      enable: true,
      withHeader: true,
      withFormat: true,
    },
    brushSelection: {
      dataCell: true,
      rowCell: true,
      colCell: true,
    },
    multiSelection: true,
  },
};

describe('PivotSheet Special Dimension Values Copy Tests', () => {
  let s2: SpreadSheet;

  beforeEach(async () => {
    s2 = createPivotSheet(s2Options);
    s2.setDataCfg(mockDataConfig);
    await s2.render();
  });

  test('should correctly copy data with header', () => {
    const { rowLeafNodes, colLeafNodes } = s2.facet.getLayoutResult();
    const cells = s2.facet.getDataCells().map((cell) => {
      const meta = cell.getMeta();
      const colId = String(colLeafNodes[meta.colIndex].id);
      const rowId = isEmpty(rowLeafNodes)
        ? String(meta.rowIndex)
        : String(rowLeafNodes[meta.rowIndex].id);

      return { ...meta, colId, rowId };
    });

    s2.interaction.changeState({
      cells,
      stateName: InteractionStateName.SELECTED,
    });

    const data = getSelectedData(s2);

    expect(data).toMatchSnapshot();
  });
});
