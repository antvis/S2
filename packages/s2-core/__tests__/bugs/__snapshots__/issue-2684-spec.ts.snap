// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PivotSheet Special Dimension Values Copy Tests should correctly copy data with header 1`] = `
Array [
  Object {
    "content": "	两片罐事业部	宝翼制罐	河北制罐	成都制罐	哈尔滨制罐
	dq	dq	dq	dq	dq
2018年-三月	53332				
2019年-三月	53332				
2023年-三月	53332	67456	53332	66293.333333	97192.090909
2023年-四月	67454	0	70442.733333	56665	100535.760606
2023年-五月	23566	-10000	73786.40303	65554	103879.430303
2023年-六月	67456	22222	77130.072727	53332	107223.1
2023年-七月	0	39998	80473.742424	70442.733333	61292.833333
2023年-八月	0	48877	83817.412121	73786.40303	66293.333333
2023年-九月	22222	49999	87161.081818	77130.072727	56665",
    "type": "text/plain",
  },
  Object {
    "content": "<meta charset=\\"utf-8\\"><table><tbody><tr><td></td><td>两片罐事业部</td><td>宝翼制罐</td><td>河北制罐</td><td>成都制罐</td><td>哈尔滨制罐</td></tr><tr><td></td><td>dq</td><td>dq</td><td>dq</td><td>dq</td><td>dq</td></tr><tr><td>2018年-三月</td><td>53332</td><td></td><td></td><td></td><td></td></tr><tr><td>2019年-三月</td><td>53332</td><td></td><td></td><td></td><td></td></tr><tr><td>2023年-三月</td><td>53332</td><td>67456</td><td>53332</td><td>66293.333333</td><td>97192.090909</td></tr><tr><td>2023年-四月</td><td>67454</td><td>0</td><td>70442.733333</td><td>56665</td><td>100535.760606</td></tr><tr><td>2023年-五月</td><td>23566</td><td>-10000</td><td>73786.40303</td><td>65554</td><td>103879.430303</td></tr><tr><td>2023年-六月</td><td>67456</td><td>22222</td><td>77130.072727</td><td>53332</td><td>107223.1</td></tr><tr><td>2023年-七月</td><td>0</td><td>39998</td><td>80473.742424</td><td>70442.733333</td><td>61292.833333</td></tr><tr><td>2023年-八月</td><td>0</td><td>48877</td><td>83817.412121</td><td>73786.40303</td><td>66293.333333</td></tr><tr><td>2023年-九月</td><td>22222</td><td>49999</td><td>87161.081818</td><td>77130.072727</td><td>56665</td></tr></tbody></table>",
    "type": "text/html",
  },
]
`;
